File: app/src/main/java/com/pioneer/aaron/dolly/fork/DataBaseOperator.java
package com.pioneer.aaron.dolly.fork;

import android.app.Application;
import android.content.ContentProviderOperation;
import android.content.ContentResolver;
import android.content.ContentValues;
import android.content.Context;
import android.content.OperationApplicationException;
import android.database.Cursor;
import android.os.RemoteException;
import android.provider.CallLog;
import android.provider.ContactsContract;

import com.pioneer.aaron.dolly.fork.calllog.ForkCallLogData;
import com.pioneer.aaron.dolly.utils.Matrix;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

/**
 * Created by Aaron on 4/27/17.
 */

public class DataBaseOperator {
    private static final String TAG = "aaron";

    public static final String CALLLOG_ENCRYPT = "encrypt_call";
    public static final String CALLLOG_FEATURE = "features";
    public static final String CALLLOG_CALL_TYPE = "call_type";
    public static final String CALLLOG_IS_PRIMARY = "is_primary";
    public static final String CALLLOG_SUBJECT = "subject";
    public static final String CALLLOG_POST_CALL_TEXT = "post_call_text";

    public final static String[] CALLLOG_COLUMNS = new String[]{
            CALLLOG_CALL_TYPE,
            CALLLOG_ENCRYPT,
            CALLLOG_FEATURE,
            CALLLOG_IS_PRIMARY,
            CALLLOG_SUBJECT,
            CALLLOG_POST_CALL_TEXT
    };

    private static volatile DataBaseOperator INSTANCE;
    private Context mContext;
    private ExecutorService mSingleExecutorService;
    private HashMap<String, Boolean> mColumnExists;

    private DataBaseOperator(Context context) {
        mContext = context;
        mSingleExecutorService = Executors.newSingleThreadExecutor();
        mSingleExecutorService.execute(mCheckColumnsRunnable);
    }

    /**
     * @param context {@link Application#getApplicationContext()} ONLY!!!
     * @return
     */
    public static DataBaseOperator getInstance(Context context) {
        if (INSTANCE == null) {
            synchronized (DataBaseOperator.class) {
                if (INSTANCE == null) {
                    INSTANCE = new DataBaseOperator(context);
                }
            }
        }
        return INSTANCE;
    }

    private Runnable mCheckColumnsRunnable = new Runnable() {
        @Override
        public void run() {
            mColumnExists = checkColumnsExists(DataBaseOperator.CALLLOG_COLUMNS);
        }
    };

    private HashMap<String, Boolean> checkColumnsExists(String... columns) {
        mColumnExists = new HashMap<>();
        HashMap<String, Boolean> hashMap = new HashMap<>();
        Cursor cursor = null;
        ContentResolver contentResolver = mContext.getContentResolver();
        cursor = contentResolver.query(CallLog.Calls.CONTENT_URI, null, null, null, null, null);

        if (cursor != null) {
            for (String column : columns) {
                hashMap.put(column, cursor.getColumnIndex(column) != -1);
            }
        }
        mColumnExists = hashMap;
        return hashMap;
    }

    public HashMap<String, Boolean> getColumnsExists() {
        if (mColumnExists == null || mColumnExists.isEmpty()) {
            return checkColumnsExists(DataBaseOperator.CALLLOG_COLUMNS);
        } else {
            return mColumnExists;
        }
    }
}

File: app/src/main/java/com/pioneer/aaron/dolly/fork/DataBaseOperator.kt
package com.pioneer.aaron.dolly.fork

import android.app.Application
import android.content.Context
import android.database.Cursor
import android.provider.CallLog
import java.lang.ref.WeakReference
import java.util.*
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors

/**
 * Created by Aaron on 4/27/17.
 */

class DataBaseOperator private constructor(mContext: Context) {
    private val mSingleExecutorService: ExecutorService = Executors.newSingleThreadExecutor()
    private var mColumnExists: HashMap<String, Boolean>? = null

    private val mCheckColumnsRunnable = Runnable { mColumnExists = checkColumnsExists(*DataBaseOperator.CALLLOG_COLUMNS) }

    val columnsExists: HashMap<String, Boolean>
        get() = if (mColumnExists == null || mColumnExists!!.isEmpty()) {
            checkColumnsExists(*DataBaseOperator.CALLLOG_COLUMNS)
        } else {
            mColumnExists!!
        }

    private var mWeakRef: WeakReference<Context>

    init {
        mSingleExecutorService.execute(mCheckColumnsRunnable)
        mWeakRef = WeakReference(mContext)
    }

    private fun checkColumnsExists(vararg columns: String): HashMap<String, Boolean> {
        mColumnExists = HashMap()
        val hashMap = HashMap<String, Boolean>()
        val cursor: Cursor?
        if (mWeakRef.get() == null) {
            return hashMap
        }
        val contentResolver = mWeakRef.get()!!.contentResolver
        cursor = contentResolver.query(CallLog.Calls.CONTENT_URI, null, null, null, null, null)

        if (cursor != null) {
            for (column in columns) {
                hashMap[column] = cursor.getColumnIndex(column) != -1
            }
            cursor.close()
        }
        mColumnExists = hashMap
        return hashMap
    }

    companion object {
        private val TAG = "DataBaseOperator"

        const val CALLLOG_ENCRYPT = "encrypt_call"
        const val CALLLOG_FEATURE = "features"
        const val CALLLOG_CALL_TYPE = "call_type"
        const val CALLLOG_IS_PRIMARY = "is_primary"
        const val CALLLOG_SUBJECT = "subject"
        const val CALLLOG_POST_CALL_TEXT = "post_call_text"

        val CALLLOG_COLUMNS = arrayOf(CALLLOG_CALL_TYPE, CALLLOG_ENCRYPT, CALLLOG_FEATURE, CALLLOG_IS_PRIMARY, CALLLOG_SUBJECT, CALLLOG_POST_CALL_TEXT)

        @Volatile
        private var INSTANCE: DataBaseOperator? = null

        /**
         * @param context [Application.getApplicationContext] ONLY!!!
         * @return
         */
        fun getInstance(context: Context): DataBaseOperator {
            if (INSTANCE == null) {
                synchronized(DataBaseOperator::class.java) {
                    if (INSTANCE == null) {
                        INSTANCE = DataBaseOperator(context)
                    }
                }
            }
            return INSTANCE!!
        }
    }
}
